tfhe/src/integer/gpu/mod.rs

```diff
@@ -273,6 +274,152 @@ pub unsafe fn unchecked_scalar_mul_integer_radix_kb_async<T: UnsignedInteger, B:
     );
 }
 
+#[allow(clippy::too_many_arguments)]
+/// # Safety
+///
+/// - [CudaStreams::synchronize] __must__ be called after this function as soon as synchronization
+///   is required
+pub unsafe fn compress_integer_radix_async<T: UnsignedInteger>(
+    streams: &CudaStreams,
+    glwe_array_out: &mut CudaVec<T>,
+    lwe_array_in: &CudaVec<T>,
+    fp_keyswitch_key: &CudaVec<u64>,
+    message_modulus: MessageModulus,
+    carry_modulus: CarryModulus,
+    compression_glwe_dimension: GlweDimension,
+    compression_polynomial_size: PolynomialSize,
+    lwe_dimension: LweDimension,
+    ks_base_log: DecompositionBaseLog,
+    ks_level: DecompositionLevelCount,
+    lwe_per_glwe: u32,
+    storage_log_modulus: u32,
+    num_blocks: u32,
+) {
+    assert_eq!(
+        streams.gpu_indexes[0],
+        lwe_array_in.gpu_index(0),
+        "GPU error: all data should reside on the same GPU."
+    );
```

Comment:
not sure I understand what the streams.gpu_indexes[0] and lwe_array_in.gpu_index(0) mean

Reply:
That check comes from GPU's keyswitch entry point. In single-gpu runs we assume that initially `lwe_array_in` resides in the GPU which id is in `streams.gpu_indexes[0]` (which may or may not be `0`).

Reply:
what does the gpu_index(0) mean in that case ? which GPU the lwe at index 0 is in ?

Reply:
I guess one thing I don't understand is that the vec needs to be of length 1 ? otherwise all gpu indices need to be checked right ? here we are checking only the first element of an array ?

Reply:
We currently don't use (and don't have plans to use, afaik) other positions in `lwe_array_in.gpu_index()` than 0. So we only need to check it.